File: app/src/main/java/com/github/colinjeremie/willyoufindit/utils/FilterGenreListTask.java
package com.github.colinjeremie.willyoufindit.utils;

import android.os.AsyncTask;

import com.deezer.sdk.model.Genre;

import java.util.ArrayList;
import java.util.List;

/**
 * * WillYouFindIt
 * Created by jerem_000 on 4/23/2016.
 */
public class FilterGenreListTask extends AsyncTask<String, Void, List<Genre>> {
    private final OnListFilteredListener mCallBack;
    private final List<Genre> mOriginalValues;

    public FilterGenreListTask(List<Genre> pValues, OnListFilteredListener pCallback){
        this.mOriginalValues = pValues;
        this.mCallBack = pCallback;
    }

    @Override
    protected List<Genre> doInBackground(String ... params) {
        if (params.length > 0){
            String search = NormalizeStr.normalizeIt(params[0]).toLowerCase();
            List<Genre> valuesFiltered = new ArrayList<>();

            for (Genre tmp : mOriginalValues){
                String title = NormalizeStr.normalizeIt(tmp.getName()).toLowerCase();

                if (title.contains(search)){
                    valuesFiltered.add(tmp);
                }
            }
            return valuesFiltered;
        }
        return mOriginalValues;
    }

    @Override
    protected void onPostExecute(List<Genre> Genres) {
        super.onPostExecute(Genres);
        if (mCallBack != null){
            mCallBack.onListFiltered(Genres);
        }
    }

    public interface OnListFilteredListener{
        void onListFiltered(List<Genre> pList);
    }
}

File: app/src/main/java/com/github/colinjeremie/willyoufindit/utils/FilterGenreListTask.kt
package com.github.colinjeremie.willyoufindit.utils

import android.os.AsyncTask

import com.deezer.sdk.model.Genre

import java.util.ArrayList

class FilterGenreListTask(private val originalValues: List<Genre>,
                          private val callBack: OnListFilteredListener?) : AsyncTask<String, Void, MutableList<Genre>>() {

    override fun doInBackground(vararg params: String): MutableList<Genre> {
        if (params.isNotEmpty()) {
            val search = params[0].normalize().toLowerCase()
            val valuesFiltered = mutableListOf<Genre>()

            for (tmp in originalValues) {
                val title = tmp.name.normalize().toLowerCase()

                if (title.contains(search)) {
                    valuesFiltered.add(tmp)
                }
            }
            return valuesFiltered
        }
        return originalValues.toMutableList()
    }

    override fun onPostExecute(genres: MutableList<Genre>) {
        super.onPostExecute(genres)
        callBack?.onListFiltered(genres)
    }

    interface OnListFilteredListener {
        fun onListFiltered(list: MutableList<Genre>)
    }
}
